Fix outer traversal to use lexical chain

smarr opened the pull request:
TODO: implement fix @richard-roberts, this is currently only the test documenting the broken behavior. (because of laziness the PR still contains elements of the release preparation)

This test demonstrates that our traversal of outer objects currently
does follow the bottom of the chain of outer objects, i.e., the outer
object of the most concrete class, instead of following the lexical chain
defined by the lexical scopes.

A solution to this issue probable needs to adapt the outer object read to take the full and correct lexical chain into account.
Roughly something like this:

```java
  @ExplodeLoop
  private Object getEnclosingObject(final SClass lexicalClass) {
    int ctxLevel = contextLevel - 1; // 0 is already covered with specialization
    SObjectWithClass enclosing = lexicalClass.getEnclosingObject();

    while (ctxLevel > 0) {
      ctxLevel--;
      
      SClass cls = enclosing.getSOMClass();
      // TODO: cls = lexicalClass(cls, ctxMixinId[ctx]);
      
      enclosing = cls.getEnclosingObject();
    }
    return enclosingObj.profile(enclosing);
  }
```

smarr committed:
Document broken outer traversal with a test

This test demonstrates that our traversal of outer objects currently
does follow the bottom of the chain of outer objects, i.e., the outer
object of the most concrete class, instead of following the lexical chain
defined by the lexical scopes.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented:
@richard-roberts the simplest solution to this issue might be to construct a chain of outer object read nodes. this might even simplify the outer object read node itself. So, each node, steps out exactly once, and then relies on the next one to finish it correctly.

ghost commented:
@smarr Yes I agree, replacing the logic with a chain of one-step outerReadNodes seems like the cleanest solution. I will prototype it on ASAP and push here.

Commit:
Fix outer traversal to use lexical instead of dynamic outer chain.

- removed usage of contextLevel, because it is not precise enough
- changed lookupSlotOrClass to return list of mixin ids
- adapted resolution of implict sends to use mixin id list
- adapted MethodBuilder's logic for building outerObjectReads
- added explicit OOR.computeOuter()
  - this avoids reexecution of the receiver expression
    Generally, it should be safe already, but we don’t want to risk any
    side-effectful reexecution.

Co-authored-by: Stefan Marr <git@stefan-marr.de>
Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented:
@richard-roberts thanks a lot for fixing this!
The solution looks very good.

To keep the history relevant, I squashed most things into one commit.

I also pushed a few extra changes.
The idea is that we avoid reversing the list of outer ids twice.
The traversal is naturally from the most inner scope to the outer scope. With this in mind, we do not need to reverse the id list twice (once by inserting them into the list at the front, a second time by using a reverse loop).

Going to try to construct a few more tests (I also changed the comment on the existing test, at least one bit was an absolute statement ("can follow any lexical class"), think the test isn't exhaustive enough to justify "any", just yet).

smarr commented on src/som/interpreter/nodes/ResolvingImplicitReceiverSend.java:
```diff
@@ -73,38 +73,49 @@ public Object doPreEvaluated(final VirtualFrame frame, final Object[] args) {
+      for (int i = result.size() - 2; i > -1; i--) {
+        enclosingMixin = result.get(i);
+
+        newReceiverNodeForOuterSend =
```
This update is problematic, because it is a transitory update, which is not leading to a correct lookup, but is visible to other threads. Will fix this.

smarr commented on src/som/interpreter/nodes/ResolvingImplicitReceiverSend.java:
```diff
@@ -73,38 +73,49 @@ public Object doPreEvaluated(final VirtualFrame frame, final Object[] args) {
       if (newReceiverNodeForOuterSend != null) {
         // need to recalculate the real receiver for outer sends
-        args[0] = newReceiverNodeForOuterSend.executeEvaluated(args[0]);
+        args[0] = newReceiverNodeForOuterSend.executeGeneric(frame);
```
@richard-roberts While I agree that we need this semantics, the change itself is highly problematic, because it relies on complex assumptions to be true.
Generally, we can not reevaluate any nodes arbitrarily, because they might have side effects that are visible on the language level.

Here the key assumption is that the evaluation of the outer chain won't cause side effects.
Specifically, the evaluation of the original `args[0]` is not allowed to have side effects, so that a reevaluation is unproblematic. I have been thinking of ensuring this with assertions, which seems hard. So, I'll put in explicit code that guarantees this.

smarr committed:
Restructure RIRS.specialize to be race free

Carefully group all accesses to replacedBy and newReceiverNodeForOuterSend.
Technically, this could also be realized with volatile fields, but I wanted to have the more clear grouping with explicit synchronized, just to be sure.

The reordering makes sure all racy fields of the node are read and written at most once. This is to prevent any interference from other threads having problematic effects.

Signed-off-by: Stefan Marr <git@stefan-marr.de>

smarr commented:
Ok, I'll assume that there is nothing contentious here.
So, will cleanup the commits and then look into merging it.

smarr commented:
I added a bit more testing, and I think I am happy with merging once tests/benchmarks completed.

ghost commented:
Thanks for cleaning up the implementation and the comments you added. Looks good to me!
